### DISCUSSION TITLE
Proposal: stringof operator

### DISCUSSION THREAD

**Discussion by Flithor** [dotnet/csharplang#5520] [General]
As we know that `nameof()` always be requeired by some people to add some new behavior to make it can convert strings that contains more info.
But... how about a new one?
`stringof()`
> Or this purpose is clearer one: [syntax symbol string](https://github.com/dotnet/csharplang/discussions/9936)

The current design being considered is that any name that is valid in `nameof` will be allowed.
It will compile the name in code within the parentheses directly into a constant string.

The examples:

Expression | Value
-- | --
stringof(System.Collections.Generic.List<>) | `"System.Collections.Generic.List<>"`
stringof(global::System.Func\<\>) | `"global::System.Func<>"`
stringof(ImmutableArray\<\>.Builder.Capacity) | `"ImmutableArray<>.Builder.Capacity"`<br/>Suggest quick fix: To full name ↓
stringof(System.Collections.ImmutableArray<>.Builder.Capacity) | `"System.Collections.ImmutableArray<>.Builder.Capacity"`
stringof(List\<int\>) | `"List<int>"`<br/>suggest quick fix: To full CLR type name ↓
stringof(List<System.Int32>) | `"List<System.Int32>"`
stringof(TSomeTypeParameterInScope) | `"TSomeTypeParameterInScope"`
stringof((int A, int B)) | `"(int A, int B)"`
stringof(int[]) | `"int[]"`
stringof(int*) | `"int*"`
stringof(int?) | `"int?"`
stringof(string?) | `"string?"`

However, although `nameof` allows new line, considering the intended function of this new operator, the parentheses in `stringof` may not allow line break chars.

Expression | Value
-- | --
**nameof**<br/>(System.Collections.<br/>Generic.List\<string>)| **nameof** returns:<br/>`"List"`
stringof<br/>(System.Collections.<br/>Generic.List\<string>)| ❌ Compile Error

**Comment by svick** [comment]
This is quite similar to https://github.com/dotnet/csharplang/discussions/701.

**Comment by CyrusNajmabadi** [comment]
Moving to csharplang.

**Comment by colejohnson66** [comment]
I would greatly appreciate something like this. When writing obsoletion messages, I like to use `nameof`. But that forces (assuming I do use `nameof`) spelling out the various segments multiple times. Especially with nested class/property chains:

```csharp
[Obsolete($"Use {nameof(TheLookup)}.{nameof(TheLookup.Entry)}.{nameof(TheLookup.Entry.TheValue)}.")]
```

This comes from:

```csharp
public sealed class TheLookup
{
    public readonly struct Entry
    {
        public Thing TheValue { get; }
    }
}
```

Ideally this would just be:

```csharp
[Obsolete($"Use {stringof(TheLookup.Entry.TheValue)}.")]
```

I'm sure others can attest to this pain.

I'm not sure how this would work w/r/t refactoring, but something that is a compile-time constant and is guaranteed to be an existing identifier (like `nameof`) would be much appreciated.

**Reply by AndersKehlet** [reply]
If we had extension constants I think we could solve this in a reasonably ergonomic way with a source generator.
You'd write something like `TheLookup.Entry.fullnameof`, and the generator would pick up on the invoked member name and emit an extension const on the type. Including a member name (`TheValue`) would require an extra step, but this would still reduce the number of operations from arbitrarily large, to at most 2.
